Mozc Renderer: Windows are not placed under the cursor with HiDPI display

## Description
Qt6 does not support AA_EnableHighDpiScaling as mentioned in compiler warnings.

```
renderer/qt/qt_server.cc:109:34: warning: 'Qt::AA_EnableHighDpiScaling' is deprecated: High-DPI scaling is always enabled.
This attribute no longer has any effect. [-Wdeprecated-declarations]
 109 |   QApplication::setAttribute(Qt::AA_EnableHighDpiScaling);
```

It now always uses virtual pixels. However, IBus only knows about device pixels and it sends a preedit rect in device pixel coordinates. Thus, a candidate window is usually placed bottom-right of the appropriate place.

## Steps to reproduce
Steps to reproduce the behavior:
1. Set display DPI to 120
2. Open gEdit, type some keys
3. The suggestion window and candidate window is not shown in a correct place

## Screenshots
![Screenshot_20231017_222954](https://github.com/google/mozc/assets/160235/34764de2-8fb2-4d72-b356-57a56035505c)

## Version or commit-id
Mozc-2.29.5220.102+24.11.oss

## Environment
 - OS: openSUSE Leap 15.5, 
 - IMF (for Linux): IBus 1.5.25
 - Related Applications
   * Qt 6.4.2
   * KDE Plasma 5.27.4.1-bp155.1.7

## Investigations
 - Whether this issue happens on both Ibus and other IMF (e.g. Fcitx, uim).
   + N/A
 - Whether this issue happens on other IMEs (e.g. Anthy, SKK).
   + N/A
 - What applications this issue happens on (e.g. Chromium, gedit).
   + gEdit, VSCode
 - What applications this issue does not happen on (e.g. Chromium, gedit).
 - (optional) What versions or commit-ids this issue did not happen on
   (e.g. Mozc-2.28.4960.100+24.11.oss).
   + list of versions or commit-ids, or no-investigation.

## Additional context

* I cannot find any function that converts device pixels to virtual. We might need to talk with the Qt community.
* Here is Qt-side code that converts virtual to device   
   https://codereview.qt-project.org/c/qt/qtbase/+/451617/2/src/plugins/platforminputcontexts/ibus/qibusplatforminputcontext.cpp
* Workaround: `QT_ENABLE_HIGHDPI_SCALING=0 /usr/lib64/mozc/mozc_renderer`
   * https://doc.qt.io/qt-6/highdpi.html

--- a/src/renderer/qt/qt_window_manager.cc
+++ b/src/renderer/qt/qt_window_manager.cc
@@ -339,13 +339,69 @@ void FillCandidates(const commands::Candidates &candidates,
   const int width = kColumn0Width + max_width1 + max_width2 + kColumn3Width;
   table->resize(width, total_height);
 }
+
+class VirtualRect {
+ public:
+  static VirtualRect FromNativeRect(const Rect &native_rect) {
+    for (const QScreen *screen : QGuiApplication::screens()) {
+      const Rect rect = TranslateToVirtual(screen, native_rect);
+      const QRect screen_rect = screen->geometry();
+
+      // Use top left to locate a screen
+      if (screen_rect.contains(rect.Left(), rect.Top())) {
+        return VirtualRect(rect, mozc::renderer::GetRect(screen->geometry()));
+      }
+    }
+
+    // fall back to primary screen
+    // TODO: Return the nearest monitor rect instead.
+    // See GetMonitorRect
+    const QScreen *screen = QGuiApplication::primaryScreen();
+    const Rect rect = TranslateToVirtual(screen, native_rect);
+    return VirtualRect(rect, mozc::renderer::GetRect(screen->geometry()));
+  }
+
+  Rect GetRect() const { return rect_; }
+
+  Rect GetMonitorRect() const { return monitor_rect_; }
+
+ private:
+  VirtualRect(const Rect &rect, const Rect &monitor_rect)
+      : rect_(rect), monitor_rect_(monitor_rect) {}
+
+  static Rect TranslateToVirtual(const QScreen *screen,
+                                 const Rect &native_rect) {
+    const double device_pixel_ratio = screen->devicePixelRatio();
+    // screen_left, screen_top have the same value in both virtual and native
+    // coordinate
+    const int screen_left = screen->geometry().x();
+    const int screen_top = screen->geometry().y();
+    const int dx = native_rect.Left() - screen_left;
+    const int dy = native_rect.Top() - screen_top;
+    const int vx =
+        static_cast<int>(std::floor(dx / device_pixel_ratio)) + screen_left;
+    const int vy =
+        static_cast<int>(std::floor(dy / device_pixel_ratio)) + screen_top;
+
+    return Rect(vx, vy, native_rect.Width() / device_pixel_ratio,
+                native_rect.Height() / device_pixel_ratio);
+  }
+
+  Rect rect_;
+  Rect monitor_rect_;
+};
 }  // namespace
 
 Point QtWindowManager::GetWindowPosition(
     const commands::RendererCommand &command, const Size &win_size) {
-  const Rect preedit_rect = GetRect(command.preedit_rectangle());
+  const Rect native_preedit_rect = GetRect(command.preedit_rectangle());
+  // Qt6 applications use virtual coordinates. Since IBus uses the device-pixel
+  // native coordinate system, we need to translate a received rect to virtual.
+  const VirtualRect virtual_rect =
+      VirtualRect::FromNativeRect(native_preedit_rect);
+  const Rect preedit_rect = virtual_rect.GetRect();
   const Point win_pos = Point(preedit_rect.Left(), preedit_rect.Bottom());
-  const Rect monitor_rect = GetMonitorRect(win_pos.x, win_pos.y);
+  const Rect monitor_rect = virtual_rect.GetMonitorRect();
   const Point offset_to_column1(kColumn0Width, 0);
 
   const Rect adjusted_win_geometry =

